### DISCUSSION TITLE
Testing a large database > 20k warehouses

### DISCUSSION THREAD

**Discussion by JP-Reddy** [TPC-Council/HammerDB#457] [Q&A] [answered]
We're trying to run Hammerdb on an Azure Sql db with over 20k warehouses but we want to understand few things.

To get a balanced and distributed workload, 
1) Should we be using around 5k VUs(~1/4th of warehouses) like you mentioned in the docs for the workload to be more distributed? Using 5k VUs would need either a really powerful VM or multiple VMs to be running together. 
2) How about using 5k VUs but with async scaling with key and thinking time?
3) Would it be better to use maybe lesser VUs and make it use all warehouses?
4) Is there a different approach which we should be taking for such databases?

We would like to get your perspective on this.

**Comment by sm-shaw** [comment, marked as answer]
Best place to start for this question is this blog post and the section cached vs scaled workloads https://www.hammerdb.com/blog/uncategorized/hammerdb-for-managers/.  

Note that with a cached workload the option use all warehouses can increase the I/O by using the available dataset.

With 20k warehouses you are not going to run a cached workload, so you need to run a scaled one with the option asynchronous scaling. The scaled workload is recommended for advanced users.  When doing this each virtual user manages multiple database sessions, these sessions using keying and thinking time like sleep time that is managed asynchronously, and the virtual user processes the transactions for each when they wake up. 

When doing this, it is likely that you will use some form of middleware, so your large number of clients connect to the middleware that then consolidates these connections into a smaller number that actually connect directly into the database. (This is how an official TPC-C benchmark would be run). 

With the scaled workload the keying and thinking means that you are running approximately 1 NOPM per session, so for example 100 virtual users with 100 connections will give you 10,000 NOPM, 200=20,000 NOPM and so on. Therefore, you don't want more sessions per VU than you would expect NOPM from a single VU running in default mode and I would recommend sizing below this. It is the keying and thinking time that is managed asynchronously so ensuring that the sessions are sleeping adequately to scale the load is important .i.e. it would not work if the keying and thinking time was 0 as 1 session would be enough to drive the virtual user, and then we are back to the cached workload! (this point about the scalability of asynchronous connection is often misunderstood). 

You may also need to configure multiple clients to generate the size of load you need, so as noted this functionality is for advanced users.
